prototypes/matrix_based_advection_diffusion/matrix_based_advection_diffusion.cc

```diff
@@ -0,0 +1,1910 @@
+  std::vector<Tensor<1, dim>> advection_term_values(n_q_points);
+
+  for (unsigned int level = 0; level < triangulation.n_global_levels(); ++level)
+    {
+      const IndexSet dof_set =
+        DoFTools::extract_locally_relevant_level_dofs(dof_handler, level);
+      boundary_constraints[level].reinit(dof_set);
+      boundary_constraints[level].add_lines(
+        mg_constrained_dofs.get_refinement_edge_indices(level));
+      boundary_constraints[level].add_lines(
+        mg_constrained_dofs.get_boundary_indices(level));
+
+      boundary_constraints[level].close();
+    }
+
+  for (const auto &cell : dof_handler.cell_iterators())
+    if (cell->level_subdomain_id() == triangulation.locally_owned_subdomain())
+      {
+        cell_matrix = 0;
+
+        fe_values.reinit(cell);
+
+        fe_values.get_function_values(solution, newton_step_values);
+
+        advection_field.value_list(fe_values.get_quadrature_points(),
+                                   advection_term_values);
+
+        for (unsigned int q = 0; q < n_q_points; ++q)
+          {
+            const double dx = fe_values.JxW(q);
+            const double nonlinearity =
+              (parameters.nonlinearity ?
+                 1e-01 * std::exp(newton_step_values[q]) :
+                 0.0);
+
+            for (unsigned int i = 0; i < dofs_per_cell; ++i)
+              {
+                const double         phi_i      = fe_values.shape_value(i, q);
+                const Tensor<1, dim> grad_phi_i = fe_values.shape_grad(i, q);
+
+                for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                  {
+                    const double         phi_j = fe_values.shape_value(j, q);
+                    const Tensor<1, dim> grad_phi_j =
+                      fe_values.shape_grad(j, q);
+
+                    cell_matrix(i, j) +=
+                      (1 / parameters.peclet_number * grad_phi_i * grad_phi_j +
+                       advection_term_values[q] * grad_phi_j * phi_i -
+                       phi_i * nonlinearity * phi_j) *
+                      dx;
+
+                    if (parameters.stabilization)
+                      {
+                        double h = 0;
+                        if (dim == 2)
+                          {
+                            h = std::sqrt(4. * cell->measure() / M_PI) /
+                                parameters.element_order;
+                          }
+                        else if (dim == 3)
+                          {
+                            h = std::pow(6 * cell->measure() / M_PI, 1. / 3.) /
+                                parameters.element_order;
+                          }
+
+                        double tau = std::pow(
+                          std::pow(1 / (parameters.peclet_number * h * h), 2) +
+                            std::pow(2 * advection_term_values[q].norm() / h,
+                                     2),
+                          -0.5);
```

Comment:
not important here, but it's better to use utilities::fixed_int_power than std::pow for fixed integer power.
https://www.dealii.org/current/doxygen/deal.II/namespaceUtilities.html#a834c6b595ada6f8e73c78d9bbd8ec551